File: ios/Sources/CapacitorPosthogPlugin/CapacitorPosthogPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CapacitorPosthogPlugin)
public class CapacitorPosthogPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CapacitorPosthogPlugin"
    public let jsName = "CapacitorPosthog"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "startSessionReplay", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startSessionReplaySession", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "endSessionReplay", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isSessionReplayEnabled", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "identifySessionReplay", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = CapacitorPosthog()

    @objc func startSessionReplay(_ call: CAPPluginCall) {
        let sessionId = call.getString("sessionId") ?? ""
        let sdkOptions = call.getObject("sdkOptions") ?? [:]
        let sdkReplayConfig = call.getObject("sdkReplayConfig") ?? [:]
        let decideReplayConfig = call.getObject("decideReplayConfig") ?? [:]

        implementation.startSessionReplay(
            sessionId: sessionId,
            sdkOptions: sdkOptions,
            sdkReplayConfig: sdkReplayConfig,
            decideReplayConfig: decideReplayConfig
        ) { result in
            call.resolve()
        }
    }

    @objc func startSessionReplaySession(_ call: CAPPluginCall) {
        let sessionId = call.getString("sessionId") ?? ""

        implementation.startSessionReplaySession(sessionId: sessionId) { result in
            call.resolve()
        }
    }

    @objc func endSessionReplay(_ call: CAPPluginCall) {
        implementation.endSessionReplay { result in
            call.resolve()
        }
    }

    @objc func isSessionReplayEnabled(_ call: CAPPluginCall) {
        implementation.isSessionReplayEnabled { result in
            call.resolve(["value": result])
        }
    }

    @objc func identifySessionReplay(_ call: CAPPluginCall) {
        let distinctId = call.getString("distinctId") ?? ""
        let anonymousId = call.getString("anonymousId") ?? ""

        implementation.identifySessionReplay(
            distinctId: distinctId,
            anonymousId: anonymousId
        ) { result in
            call.resolve()
        }
    }
}

File: android/src/main/java/com/rediska1114/plugins/posthog/CapacitorPosthogPlugin.java
package com.rediska1114.plugins.posthog;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "CapacitorPosthog")
public class CapacitorPosthogPlugin extends Plugin {

    private CapacitorPosthog implementation = new CapacitorPosthog();

    @PluginMethod
    public void startSessionReplay(PluginCall call) {
        String sessionId = call.getString("sessionId", "");
        JSObject sdkOptions = call.getObject("sdkOptions", new JSObject());
        JSObject sdkReplayConfig = call.getObject("sdkReplayConfig", new JSObject());
        JSObject decideReplayConfig = call.getObject("decideReplayConfig", new JSObject());

        implementation.startSessionReplay(getContext(), sessionId, sdkOptions, sdkReplayConfig, decideReplayConfig, new CapacitorPosthog.Callback() {
            @Override
            public void onComplete() {
                call.resolve();
            }
        });
    }

    @PluginMethod
    public void startSessionReplaySession(PluginCall call) {
        String sessionId = call.getString("sessionId", "");

        implementation.startSessionReplaySession(sessionId, new CapacitorPosthog.Callback() {
            @Override
            public void onComplete() {
                call.resolve();
            }
        });
    }

    @PluginMethod
    public void endSessionReplay(PluginCall call) {
        implementation.endSessionReplay(new CapacitorPosthog.Callback() {
            @Override
            public void onComplete() {
                call.resolve();
            }
        });
    }

    @PluginMethod
    public void isSessionReplayEnabled(PluginCall call) {
        implementation.isSessionReplayEnabled(new CapacitorPosthog.BooleanCallback() {
            @Override
            public void onResult(boolean result) {
                JSObject ret = new JSObject();
                ret.put("value", result);
                call.resolve(ret);
            }
        });
    }

    @PluginMethod
    public void identifySessionReplay(PluginCall call) {
        String distinctId = call.getString("distinctId", "");
        String anonymousId = call.getString("anonymousId", "");

        implementation.identifySessionReplay(distinctId, anonymousId, new CapacitorPosthog.Callback() {
            @Override
            public void onComplete() {
                call.resolve();
            }
        });
    }
}
